File: telecom-field-engineer-system/auth-service/src/main/java/com/sts/auth/service/AuthService.java
package com.sts.auth.service;

import com.sts.auth.dto.*;
import com.sts.auth.entity.User;
import com.sts.auth.repository.UserRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import java.util.List;
import java.util.Optional;

/**
 * Authentication Service
 * 
 * Handles all business logic for user authentication and management.
 * 
 * FR1: Supports USER, ADMIN, ENGINEER roles
 * FR2/FR3: Common login with role-based response
 * FR4: Role-specific registration
 * FR5: Security question-based password recovery
 * FR8: Profile viewing and updating
 */
@Service
public class AuthService {

    @Autowired
    private UserRepository userRepository;

    /**
     * Register a new user
     * FR4: Users, Engineers, and Admins provide specific information for registration.
     * 
     * @param request registration details
     * @return success/failure message
     */
    public String register(RegisterRequest request) {
        // Check if email already exists
        if (userRepository.existsByEmail(request.getEmail())) {
            return "Email already registered!";
        }

        // Build user entity from request
        // Password is received already encrypted from frontend (Encrypt-js)
        User user = User.builder()
                .name(request.getName())
                .email(request.getEmail())
                .password(request.getPassword())
                .role(User.Role.valueOf(request.getRole().toUpperCase()))
                .phone(request.getPhone())
                .address(request.getAddress())
                .securityQuestion(request.getSecurityQuestion())
                .securityAnswer(request.getSecurityAnswer())
                .build();

        userRepository.save(user);
        return "Registration successful!";
    }

    /**
     * Authenticate a user
     * FR2: Common login page for users, engineers and admins.
     * FR3: Users authenticate before accessing ticket-raising functionality.
     * 
     * @param request login credentials (email + encrypted password)
     * @return LoginResponse with user details or error
     */
    public LoginResponse login(LoginRequest request) {
        Optional<User> userOpt = userRepository.findByEmail(request.getEmail());

        if (userOpt.isEmpty()) {
            return LoginResponse.builder()
                    .message("User not found!")
                    .build();
        }

        User user = userOpt.get();

        // Compare encrypted passwords (both encrypted with Encrypt-js on frontend)
        if (!user.getPassword().equals(request.getPassword())) {
            return LoginResponse.builder()
                    .message("Invalid credentials!")
                    .build();
        }

        return LoginResponse.builder()
                .userId(user.getUserId())
                .name(user.getName())
                .email(user.getEmail())
                .role(user.getRole().name())
                .message("Login successful!")
                .build();
    }

    /**
     * Forgot Password - Security Question Flow
     * FR5: Feature of security questions in case the user forgets login credentials.
     * 
     * Step 1: User provides email → system returns the security question
     * Step 2: User provides answer + new password → system verifies and updates
     * 
     * @param request forgot password details
     * @return success/failure message
     */
    public String forgotPassword(ForgotPasswordRequest request) {
        Optional<User> userOpt = userRepository.findByEmail(request.getEmail());

        if (userOpt.isEmpty()) {
            return "User not found with this email!";
        }

        User user = userOpt.get();

        // Verify security answer (case-insensitive comparison)
        if (!user.getSecurityAnswer().equalsIgnoreCase(request.getSecurityAnswer())) {
            return "Incorrect security answer!";
        }

        // Update password (already encrypted from frontend)
        user.setPassword(request.getNewPassword());
        userRepository.save(user);

        return "Password updated successfully!";
    }

    /**
     * Get security question for a user (first step of forgot password)
     * 
     * @param email the user's email
     * @return the security question or error message
     */
    public String getSecurityQuestion(String email) {
        Optional<User> userOpt = userRepository.findByEmail(email);
        if (userOpt.isEmpty()) {
            return "User not found!";
        }
        return userOpt.get().getSecurityQuestion();
    }

    /**
     * Get all users (FR11: Admin views user details)
     * 
     * @return list of all users
     */
    public List<User> getAllUsers() {
        return userRepository.findAll();
    }

    /**
     * Get users by role (FR11: Admin views technician details)
     * 
     * @param role the role to filter by
     * @return list of users with that role
     */
    public List<User> getUsersByRole(String role) {
        return userRepository.findByRole(User.Role.valueOf(role.toUpperCase()));
    }

    /**
     * Get user by ID
     * 
     * @param id the user ID
     * @return Optional containing the user
     */
    public Optional<User> getUserById(Long id) {
        return userRepository.findById(id);
    }

    /**
     * Update user profile (FR8)
     * 
     * @param id the user ID
     * @param request updated profile details
     * @return updated user or null if not found
     */
    public User updateProfile(Long id, RegisterRequest request) {
        Optional<User> userOpt = userRepository.findById(id);
        if (userOpt.isEmpty()) {
            return null;
        }

        User user = userOpt.get();
        if (request.getName() != null) user.setName(request.getName());
        if (request.getPhone() != null) user.setPhone(request.getPhone());
        if (request.getAddress() != null) user.setAddress(request.getAddress());
        if (request.getPassword() != null && !request.getPassword().isEmpty()) {
            user.setPassword(request.getPassword());
        }

        return userRepository.save(user);
    }
}

File: telecom-field-engineer-system/frontend-angular/src/app/services/auth.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { User, LoginRequest, LoginResponse, RegisterRequest, ForgotPasswordRequest } from '../models/user.model';

// ----------------------------------------------------
// VIVA EXPLANATION: What is AuthService?
// - Purpose: Communicates with the Authentication Microservice.
// - URL: Hits the backend running on http://localhost:8081/api/auth.
// - Responsibility: Manages registration, login, password recovery, and user session storage.
// ----------------------------------------------------
@Injectable({
  providedIn: 'root'
})
export class AuthService {

  private baseUrl = 'http://localhost:8081/api/auth';

  constructor(private http: HttpClient) {}

  // VIVA EXPLANATION: Registers a new User or Engineer in the authentication service.
  register(request: RegisterRequest): Observable<any> {
    return this.http.post(`${this.baseUrl}/register`, request);
  }

  // VIVA EXPLANATION: Authenticates credentials. Sends email and the encrypted password to the backend.
  login(request: LoginRequest): Observable<LoginResponse> {
    return this.http.post<LoginResponse>(`${this.baseUrl}/login`, request);
  }

  // VIVA EXPLANATION: Retrieves security question registered by the user using their email.
  getSecurityQuestion(email: string): Observable<any> {
    return this.http.get(`${this.baseUrl}/security-question?email=${email}`);
  }

  // VIVA EXPLANATION: Submits the security question answer and updates password if correct.
  forgotPassword(request: ForgotPasswordRequest): Observable<any> {
    return this.http.post(`${this.baseUrl}/forgot-password`, request);
  }

  // VIVA EXPLANATION: Retrieves all registered users (used by Admin views).
  getAllUsers(): Observable<User[]> {
    return this.http.get<User[]>(`${this.baseUrl}/users`);
  }

  // VIVA EXPLANATION: Retrieves users filtered by their role (e.g. USER, ADMIN, ENGINEER).
  getUsersByRole(role: string): Observable<User[]> {
    return this.http.get<User[]>(`${this.baseUrl}/users/role/${role}`);
  }

  // VIVA EXPLANATION: Retrieves a single user's detail by ID.
  getUserById(id: number): Observable<User> {
    return this.http.get<User>(`${this.baseUrl}/user/${id}`);
  }

  // VIVA EXPLANATION: Updates profile details (name, phone, address, optional password).
  updateProfile(id: number, request: any): Observable<User> {
    return this.http.put<User>(`${this.baseUrl}/user/${id}`, request);
  }

  // VIVA EXPLANATION: Saves the logged-in user's details and role in the browser's localStorage.
  saveSession(user: LoginResponse): void {
    localStorage.setItem('currentUser', JSON.stringify(user));
  }

  // VIVA EXPLANATION: Gets current logged-in user from localStorage.
  getCurrentUser(): LoginResponse | null {
    const data = localStorage.getItem('currentUser');
    return data ? JSON.parse(data) : null;
  }

  // VIVA EXPLANATION: Checks if a user is currently logged in.
  isLoggedIn(): boolean {
    return localStorage.getItem('currentUser') !== null;
  }

  // VIVA EXPLANATION: Retrieves the role of the logged-in user (USER, ADMIN, or ENGINEER) for routing/views.
  getRole(): string | null {
    const user = this.getCurrentUser();
    return user ? user.role : null;
  }

  // VIVA EXPLANATION: Logs out the user by clearing their session data from localStorage.
  logout(): void {
    localStorage.removeItem('currentUser');
  }
}
